jctools-core/src/main/java/org/jctools/queues/takestrategy/ParkTakeStrategy.java

```diff
@@ -0,0 +1,42 @@
+package org.jctools.queues.takestrategy;
+
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.concurrent.locks.LockSupport;
+
+public class ParkTakeStrategy<E> implements TakeStrategy<E>
+{
+
+    public volatile int storeFence = 0;
+
+    private AtomicReference<Thread> t               = new AtomicReference<Thread>(null);
+
+    @Override
+    public void signal()
+    {
+        // Make sure the offer is visible before unpark
+        storeFence = 1; // store barrier
```

Comment:
I see, so you are not actually looking for a store fence, rather for a StoreLoad to prevent the load of the parked thread moving up.
I was trying to remember what the original problem looked like, but a bit brain dead this week.